### DISCUSSION TITLE
Enable ASIO driver for Windows

### DISCUSSION THREAD

**Discussion by Mikkel3** [plugdata-team/plugdata#689] [General]
For Pure Data, there is an option in settings to select the ASIO driver to use audio. Using a USB interface this allows me to use 4 channel audio in Pure Data.
The same setting is unavailable to select in PlugData, which makes that I can only use 2 channels at a time. 
The difference in availability can be seen in the screenshots:
![puredata](https://user-images.githubusercontent.com/95369461/224952116-22e9593b-27e5-41da-ad8b-a7ffe94caad0.png)
![plugdata](https://user-images.githubusercontent.com/95369461/224952716-ffedfd84-7748-4adf-87ad-8835ca38b791.png)

**Comment by sevdh** [comment]
And in case ASIO would not be an option, is there another way to use a multichannel interface with Plugdata on Windows (and output more than 2 channels)?

**Reply by timothyschoen** [reply]
I want to fix it, but I'm afraid this is a problem in JUCE. I'll investigate some more though!

**Reply by timothyschoen** [reply]
And ASIO is unfortunately not an option, there are more apps that have the same licensing problem that I do, [Audacity](https://github.com/audacity/audacity/issues/507), for example. I could provide it as a compile-time option so the user can build their own version with ASIO support, but I'm legally not allowed to add ASIO support to a binary.

**Reply by sevdh** [reply]
Great you’re looking into this. Let’s see if there is another way besides ASIO.

**Comment by Astrinvega** [comment]
Well I just learned Pd has an ASIO option! Right in front of me...

**Reply by sevdh** [reply]
Yes, PD does. The topic here is Plugdata :-)

**Reply by Astrinvega** [reply]
Thanks for that friendly warning
A fan of the Andrew Belt style of communicating with fellow humans, I see

**Reply by tomara-x** [reply]
i feel it's always good to try being mindful of how different people (especially people we're not familiar with) have different styles of communication (also tone of voice doesn't translate well in text :wink:)

**Reply by Astrinvega** [reply]
I was trying to be friendly when I said 'at my own expense' that I had totally overlooked that Pd has an ASIO option. It wasn't to talk **only** about Pd. That's the point, that Plugdata is built with/on Pd and that I too would like to see Plugdata include the ASIO option. The response to my comment seemed 'rude' and including a smiley face didn't make the comment any less 'rude'. My rejoinder involving Andrew Belt is that he is notorious for being snarky and so rude to anyone, including developers to the project he started--VCV Rack--that some of those developers have left.

Now if you or any of the folks involved with the project would rather I not get involved and comment, then that's fine. I assumed that this area of github would be as friendly as I found my interactions to be with Timothy S on the Electrosmith forum. :) ;)

**Reply by sevdh** [reply]
I meant to be friendly but it was clearly not received like that. When I was writing "The topic here is Plugdata :-)" I meant to keep this particular discussion thread focused on Plugdata and ASIO (or other possible ways of doing multichannel audio with Plugdata on Windows).
I had and have no negative intentions and I think we share the same fascination for using multichannel audio!

**Reply by Astrinvega** [reply]
thank you sevdh for your explanation.
i am here ONLY to support the Plugdata project. just the other day i was explaining on a youtube video on the chompi (which is a daisy project) on the merits of Plugdata. this is a project worth promoting.

**Comment by sadguitarius** [comment]
FYI I was able to enable ASIO by adding `JUCE_ASIO=1` to `JUCE_COMPILE_DEFINITIONS` and adding `<MY_ASIO_SDK_DIR>/common` to the include directories of the juce target in CMakeLists.txt. Would it be sensible to add this officially so that users could compile their own ASIO-enabled builds?

**Reply by b1tcr4sh** [reply]
Could you give an example of where to add the asio sdk target? I'm not largely familiar with editing cmakelists and I'm not sure where to put this. 

Thanks

**Reply by sadguitarius** [reply]
I'm away from my dev machine right now but basically, in CMakeLists.txt you'd add a line with `JUCE_ASIO=1` under the line `set(JUCE_COMPILE_DEFINITIONS` (where all the other definitions are) and another line after the line `add_library(juce STATIC)` containing `target_include_directories(juce PRIVATE <MY_ASIO_SDK_DIR>/common)`, replacing `MY_ASIO_SDK_DIR` with wherever you've put the ASIO SDK on your computer (You need to download that separately from Steinberg). That's at least the basic idea for getting it running on your machine, although not actually a robust solution since it relies on hard-coded paths.

**Comment by sevdh** [comment]
This is really great news. Thanks for trying and sharing!

I think it would indeed be nice to add this officially.

**Comment by sadguitarius** [comment]
Looks like as of this week, Steinberg properly supports GPL3 projects using the ASIO SDK! I won't be able to submit a PR for this for a few days but if anyone wants to go for it, I don't _think_ there's anything in the way of doing this from a legal standpoint anymore.

**Reply by sevdh** [reply]
Fantastic to hear that Steinberg allows GPL3 now.

**Comment by sevdh** [comment]
As Steinberg now supports GPL3 projects would this be something to include in a future update?

**Reply by timothyschoen** [reply]
Done!

**Reply by sevdh** [reply]
Thanks so much for picking this up!!!

It is really fantastic to have this included and be able to work with ASIO multichannel projects on Windows!
